crates/scheduler/src/evaluators.rs

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::sync::Arc;
+use std::time::Instant;
+
+use uuid::Uuid;
+
+use crate::tasks::TaskRequest;
+
+#[derive(PartialEq)]
+pub enum EvaluatorType {
+    /// The evaluator returns the next request in the schedule, only if the current time is greater than the last request.
+    /// E.g. `.next()` -> 1am, at 12:30am `.next()` -> 1am, at 1:00 am `.next()` -> 2am.
+    Sequential,
+    /// The evaluator returns requests based on a fixed interval, anchored from a specific time.
+    /// E.g. `.next()` -> 1am, at 12:30am `.next()` -> 1:30am, at 1:00 am `.next()` -> 2:00am.
+    /// Resetting a timed evaluator should reset the anchor time to the current time.
+    Timed,
+    /// The evaluator returns interrupt requests.
+    /// Interrupts have the highest priority and should be delivered immediately.
+    /// The scheduler should abort any in-progress evaluation wait timers and deliver the interrupt request.
+    Interrupt,
+}
+
+pub trait Evaluator: Send + Sync {
+    /// Returns a unique identifier for this evaluator.
+    fn id(&self) -> Arc<Uuid>;
+    /// Returns the type of this evaluator.
+    fn evaluator_type(&self) -> EvaluatorType;
+    /// Returns the next task request for this evaluator.
+    fn evaluate(&mut self) -> Option<Arc<TaskRequest>>;
+    /// Clears the internal state of this evaluator to reset it to its initial state.
+    fn reset(&mut self);
+}
```

Comment:
It's not very clear what you are trying to accomplish with this. Evaluator is not a great name for what I think this is, which produces a request for the next scheduled work? Also the evaluator types are weird - why does that matter for the interface, it should be an implementation detail of the specific component.

If you call evaluate multiple times in a row, does it produce the same request? i.e. is it idempotent?

Isn't this just a channel and the component itself is responsible for putting task requests on the channel when its time to execute? (i.e. the queueing is managed internally)

Reply:
A stream interface also works, because it can defer returning until it should queue

Reply:
Ah I see, this is used by an external component and the interface is basically asking "when should I execute the next task" - and its only called when the external component needs to know when the next task should execute.

But that breaks down for the "interrupt" model because the external component already has an item waiting in the queue, then it won't call this component.

I think there needs to be two interfaces, one for future tasks and one for interrupt driven tasks, rather than trying to combine them.

Reply:
Maybe that is what you were attempting with the types? But I think that is a bad model, it should just be a separate trait.